utils/service/txmgr/txmgr.go

```diff
@@ -89,315 +65,438 @@ type ETHBackend interface {
+func (m *SimpleTxManager) craftTx(ctx context.Context, candidate TxCandidate) (*types.Transaction, error) {
+	gasTipCap, basefee, err := m.suggestGasPriceCaps(ctx)
+	if err != nil {
+		m.metr.RPCError()
+		return nil, fmt.Errorf("failed to get gas price info: %w", err)
+	}
+	gasFeeCap := calcGasFeeCap(basefee, gasTipCap)
 
-	// Initialize a subcontext for the goroutines spawned in this process.
-	// The defer to cancel is done here (in reverse order of Wait) so that
-	// the goroutines can exit before blocking on the wait group.
-	ctx, cancel := context.WithCancel(ctx)
+	// Fetch the sender's nonce from the latest known block (nil `blockNumber`)
+	childCtx, cancel := context.WithTimeout(ctx, m.cfg.NetworkTimeout)
 	defer cancel()
+	nonce, err := m.backend.NonceAt(childCtx, m.cfg.From, nil)
+	if err != nil {
+		m.metr.RPCError()
+		return nil, fmt.Errorf("failed to get nonce: %w", err)
+	}
+	m.metr.RecordNonce(nonce)
 
-	sendState := NewSendState(m.SafeAbortNonceTooLowCount)
-
-	// Create a closure that will block on submitting the tx in the
-	// background, returning the first successfully mined receipt back to
-	// the main event loop via receiptChan.
-	receiptChan := make(chan *types.Receipt, 1)
-	sendTxAsync := func(tx *types.Transaction) {
-		defer wg.Done()
+	rawTx := &types.DynamicFeeTx{
+		ChainID:   m.chainID,
+		Nonce:     nonce,
+		To:        candidate.To,
+		GasTipCap: gasTipCap,
+		GasFeeCap: gasFeeCap,
+		Data:      candidate.TxData,
+	}
 
-		txHash := tx.Hash()
-		nonce := tx.Nonce()
-		gasTipCap := tx.GasTipCap()
-		gasFeeCap := tx.GasFeeCap()
-		log := m.l.New("txHash", txHash, "nonce", nonce, "gasTipCap", gasTipCap, "gasFeeCap", gasFeeCap)
-		log.Info("publishing transaction")
+	m.l.Info("creating tx", "to", rawTx.To, "from", m.cfg.From)
 
-		err := m.backend.SendTransaction(ctx, tx)
-		sendState.ProcessSendError(err)
+	// If the gas limit is set, we can use that as the gas
+	if candidate.GasLimit != 0 {
+		rawTx.Gas = candidate.GasLimit
+	} else {
+		// Calculate the intrinsic gas for the transaction
+		gas, err := m.backend.EstimateGas(ctx, ethereum.CallMsg{
+			From:      m.cfg.From,
+			To:        candidate.To,
+			GasFeeCap: gasFeeCap,
+			GasTipCap: gasTipCap,
+			Data:      rawTx.Data,
+		})
 		if err != nil {
-			if errors.Is(err, context.Canceled) {
-				return
-			}
-			if errors.Is(err, txpool.ErrAlreadyKnown) {
-				log.Info("resubmitted already known transaction")
-				return
-			}
-			log.Error("unable to publish transaction", "err", err)
-			if sendState.ShouldAbortImmediately() {
-				log.Warn("Aborting transaction submission")
-				cancel()
-			}
-			// TODO(conner): add retry?
-			return
+			return nil, fmt.Errorf("failed to estimate gas: %w", err)
 		}
+		rawTx.Gas = gas
+	}
 
-		log.Info("transaction published successfully")
+	ctx, cancel = context.WithTimeout(ctx, m.cfg.NetworkTimeout)
+	defer cancel()
+	return m.cfg.Signer(ctx, m.cfg.From, types.NewTx(rawTx))
```

Comment:
Does the code require any network cost? looks not.

Reply:
I think Signer can be a remote signer, so in case of the remote signer there exists network cost. 
(see [`SignTransaction`](https://github.com/kroma-network/kroma/blob/feat/apply-optimism-v1.0.3-batcher/utils/signer/client/client.go#L96-L110))